autoscheduler/frontend/src/redux/reducers/schedules.ts

```diff
@@ -22,18 +24,77 @@ export interface ReplaceSchedulesAction {
     type: 'REPLACE_SCHEDULES';
     schedules: Meeting[][];
 }
-export type ScheduleAction = AddScheduleAction | RemoveScheduleAction | ReplaceSchedulesAction;
+export interface SaveScheduleAction {
+    type: 'SAVE_SCHEDULE';
+    index: number;
+}
+export interface UnsaveScheduleAction {
+    type: 'UNSAVE_SCHEDULE';
+    index: number;
+}
+export type ScheduleAction = AddScheduleAction | RemoveScheduleAction
+  | ReplaceSchedulesAction | SaveScheduleAction | UnsaveScheduleAction;
+
+interface Schedules {
+  schedules: Meeting[][];
+  savedSchedules: Meeting[][];
+}
+
+const initialSchedules: Schedules = {
+  schedules: [],
+  savedSchedules: [],
+};
+
+// returns whether a list of schedules contains a schedule with the same meetings as schedule
+export function containsSchedule(schedules: Meeting[][], schedule: Meeting[]): boolean {
+  const scheduleMeetings = new Set(schedule.map((meeting) => meeting.id));
+  return schedules.some((toCompare) => {
+    const toCompareMeetings = new Set(toCompare.map((meeting) => meeting.id));
+    return scheduleMeetings.size === toCompareMeetings.size
+      && [...scheduleMeetings].every((meeting) => toCompareMeetings.has(meeting));
+  });
+}
+
+function getUniqueSchedules(...schedules: Meeting[][]): Meeting[][] {
+  const unique: Meeting[][] = [];
+  schedules.forEach((schedule) => {
+    if (!containsSchedule(unique, schedule)) unique.push(schedule);
+  });
+  return unique;
+}
 
 // reducer
-export default function meetings(state: Meeting[][] = [], action: ScheduleAction): Meeting[][] {
+function meetings(state: Schedules = initialSchedules, action: ScheduleAction): Schedules {
   switch (action.type) {
     case ADD_SCHEDULE:
-      return [...state, action.meetings];
+      return { ...state, schedules: [...state.schedules, action.schedule] };
     case REMOVE_SCHEDULE:
-      return state.slice(0, action.index).concat(state.slice(action.index + 1));
+      return {
+        ...state,
+        schedules: state.schedules.slice(0, action.index)
+          .concat(state.schedules.slice(action.index + 1)),
+      };
     case REPLACE_SCHEDULES:
-      return action.schedules;
+      return {
+        ...state,
+        schedules: getUniqueSchedules(...state.savedSchedules, ...action.schedules),
+      };
+    case SAVE_SCHEDULE:
+      return {
+        ...state,
+        savedSchedules: state.savedSchedules
+          .concat(containsSchedule(state.savedSchedules, state.schedules[action.index])
+            ? [] : [state.schedules[action.index]]),
+      };
+    case UNSAVE_SCHEDULE:
+      return {
+        ...state,
+        savedSchedules: state.savedSchedules.filter((schedule) => (
```

Comment:
Why can we use the built-in `filter` here if we have to use a custom `contains` function at line 86?

Reply:
The custom `contains` function makes it so if 2 schedules are generated with the same classes in them, only one of them will be saved instead of having 2 identical schedules (which should be a pretty rare edge case), whereas elements in `savedSchedules` are references to elements in `allSchedules`, so checking shallow equality will work